### ISSUE TITLE
 Player is not evolvable. Use a subclass of EvolvablePlayer.

### ISSUE THREAD

**Issue by Erenhub** [Axelrod-Python/Axelrod#1345]
Hi,

I would like to see how non-evolvable strategies behave in moran process compared to evolvable ones, however when i try to give players array in a mixed way i got this error. Is there a way to create such an environment?

many thanks

**Comment by marcharper** [comment]
Can you paste your code here? My guess is that you need to use the rate type of mutation, only the "atomic" type should require an EvolveablePlayer.

**Comment by Erenhub** [comment]
Of course, i'm adding it with the error as screenshot 
![errored](https://user-images.githubusercontent.com/47353633/82842371-e8df4d00-9ee1-11ea-9a38-ba20de89e481.png)

**Comment by marcharper** [comment]
A few issues:
* `axl.Cooperator` and `axl.Defector` need to be initialized `axl.Cooperator()` and `axl.Defector()`
* the mutation type has to be set to `atomic`. See the [docs here](https://axelrod.readthedocs.io/en/stable/tutorials/further_topics/evolvable_players.html)
* `EvolvablePlayer` is a generic abstract base class with no real functionality. You'll need to choose  a specific one like the [EvolvableFSMPlayer](https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/strategies/finite_state_machines.py#L128)
* I'm actually not sure that the code supports mixing of evolvable and non-evolvable players. It's not an unreasonable thing to want to do but the necessary no-op functions on the non-evolvable players may not be present.

**Comment by Erenhub** [comment]
Hi Marc,

Now i am getting ''TypeError: Player is not evolvable. Use a subclass of EvolvablePlayer.''.

code:

C = axl.Action.C
#player1=axl.Defector  i cant give a player as argument below :(
ev1=axl.EvolvableFSMPlayer(num_states=1,initial_state=0, initial_action=C) 
playerswithevolable=[]
playerswithevolable.append(ev1)
playerswithevolable.append(axl.Defector())
playerswithevolable.append(axl.Cooperator())

mpE = axl.MoranProcess(playerswithevolable,mutation_method="atomic")
populations = mpE.play()
populations

-------------------------------------------------
error: 
![errorrrrrled](https://user-images.githubusercontent.com/47353633/83073980-0a733c80-a07a-11ea-99d4-f7245e36cce9.png)

**Comment by marcharper** [comment]
That's what I expected. @drvinceknight any thoughts? Maybe we could have some kind of wrapper that makes any player generically evolvable in a trivial way for cases like this.

**Comment by drvinceknight** [comment]
Sorry for taking so long to get back to you on this.

I like your suggestion @marcharper, a wrapper that allows player to evolve in a trivial sounds good. would another approach be to add an evolve method to the base player class that does not do anything?

**Comment by marcharper** [comment]
> I like your suggestion @marcharper, a wrapper that allows player to evolve in a trivial sounds good. Would another approach be to add an evolve method to the base player class that does not do anything?

Yes we could also do that, but then it removes the distinction between Player and EvolvablePlayer. Maybe that's a good idea but I'm not sure (for testing purposes at least).

**Comment by drvinceknight** [comment]
> I'm not sure (for testing purposes at least).

I hear you. It's two options, whatever your hunch tells you is fine by me.

**Comment by marcharper** [comment]
Post #1288 I'm thinking that it's better to make all Players evolvable, if only to remove multiple inheritance from the Player classes hierarchy. I'm not sure we actually gain much by having both `Players` and `EvolvablePlayers` other than the bit that a Player is evolvable, which could more easily handled by a classifier instead.

**Comment by marcharper** [comment]
Revisiting this, one way to fix this is to
* Merge back Player and EvolvablePlayer, e.g. add generic functions to Player that implement EvolvablePlayer's functions
* Set a function / flag on each player that indicates it's evolvability -- trivial (i.e. just clones itself) or more dynamic
* Factor out parameter classes for the strategies that are non-trivially evolvable, like the dojo library does.
